File: src/endpoints/category/get_index.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token
from src.modules.mod import MOD
from . import __libs__ as category_libs
from ..article import __libs__ as article_libs
from ..json import __libs__ as json_libs

async def get_index(params: dict = {}, req = None, _db: DB = None, _check_token = True):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.GetIndex(**params)

        # checking token
        if _check_token: checking_token(req, db, use_public=True)

        # set fields
        fields = params.fields.split(',') if params.fields else None

        # set mod
        mod = MOD(params.mod or '')

        # set tag
        tags = ','.join(params.tag.split(',')) if params.tag and params.module else None

        # set where
        where = []
        if params.name:
            where.append(f'and name LIKE "%{params.name}%"')
        if params.module:
            where.append(f'and module LIKE \'{params.module}\'')
        if params.module_srl is not None:
            if params.module_srl > 0:
                where.append(f'and module_srl = {params.module_srl}')
            else:
                where.append(f'and module_srl IS NULL')

        # get total
        total = db.get_count(
            table_name=Table.CATEGORY.value,
            where=where,
        )
        if total == 0: raise Exception('No data', 204)

        # get index
        index = db.get_items(
            table_name=Table.CATEGORY.value,
            fields=fields,
            where=where,
            limit={ 'size': params.size, 'page': params.page },
            order={ 'order': params.order, 'sort': params.sort },
            unlimited=params.unlimited,
        )

        # set where for module
        where = []
        if params.q:
            match params.module:
                case category_libs.Module.NEST:
                    where.append(f'and (title LIKE "%{params.q}%" OR content LIKE "%{params.q}%")')
                case category_libs.Module.JSON:
                    where.append(f'and (name LIKE "%{params.q}%" OR description LIKE "%{params.q}%")')

        # transform items
        def transform_item(item: dict) -> dict:
            if 'srl' in item and 'module' in item:
                item_where = [ f'and category_srl = {item['srl']}' ]
                item_where.extend(where)
                # MOD / count
                if mod.check('count'):
                    if tags:
                        item_where.append(get_tag_count_query(params.module, tags))
                    match params.module:
                        case category_libs.Module.NEST:
                            item['count'] = article_libs.get_count(db, item_where)
                        case category_libs.Module.JSON:
                            item['count'] = json_libs.get_count(db, item_where)
            return item
        index = [ transform_item(item) for item in index ]

        # MOD / none
        if mod.check('none'):
            new_item = { 'name': 'none' }
            _where = [ f'and category_srl IS NULL' ]
            _where.extend(where)
            if mod.check('count'):
                if tags:
                    _where.append(get_tag_count_query(params.module, tags))
                match params.module:
                    case category_libs.Module.NEST:
                        if params.module_srl:
                            _where.append(f'and nest_srl = {params.module_srl}')
                        new_item['count'] = article_libs.get_count(db, _where)
                    case category_libs.Module.JSON:
                        new_item['count'] = json_libs.get_count(db, _where)
            index.append(new_item)

        # MOD / all
        if mod.check('all'):
            new_item = { 'name': 'all' }
            _where = []
            _where.extend(where)
            if mod.check('count'):
                if tags:
                    _where.append(get_tag_count_query(params.module, tags))
                match params.module:
                    case category_libs.Module.NEST:
                        if params.module_srl:
                            _where.append(f'and nest_srl = {params.module_srl}')
                        new_item['count'] = article_libs.get_count(db, _where)
                    case category_libs.Module.JSON:
                        new_item['count'] = json_libs.get_count(db, _where)
            index.insert(0, new_item)

        # set result
        result = output.success({
            'message': 'Complete get category index.',
            'data': {
                'total': total,
                'index': index,
            },
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

def get_tag_count_query(module: str, tag: str) -> str:
    if not module or not tag: return ''
    return f'AND json.srl IN (SELECT map_tag.module_srl FROM map_tag WHERE map_tag.module LIKE \'{module}\' AND map_tag.tag_srl IN ({tag}))'

File: src/routes/category/get-index.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import MOD from '@/classes/MOD'
import * as articleHelper from '@/routes/article/__helper'
import * as jsonHelper from '@/routes/json/__helper'
import * as helper from './__helper'
import type { CategoryModel } from './__model'

type GetIndexParams = {
  query: CategoryModel['getIndexQuery']
}

export default async function getIndex({ query }: GetIndexParams)
{
  try
  {
    // set assets
    let _table = `${DB.TABLE.CATEGORY} AS c`
    let _where: string[] = []
    let _values: ZZ = {}
    let _join: string[] = []
    const _field = query.field ? query.field.split(',') : ''
    const _tags = (query.tag && query.module) ? query.tag : ''

    // set base params
    if (query.name)
    {
      _where.push(`AND c.name LIKE '%' || $name || '%'`)
      _values['$name'] = query.name
    }
    if (query.module)
    {
      _where.push(`AND c.module LIKE $module`)
      _values['$module'] = query.module
    }
    if (query.module === helper.MODULE.NEST && query.module_srl !== undefined)
    {
      if (query.module_srl > 0)
      {
        _where.push(`AND c.module_srl = ${query.module_srl}`)
      }
      else
      {
        _where.push(`AND c.module_srl IS NULL`)
      }
    }

    // get total
    const count = helper.count({
      table: `${DB.TABLE.CATEGORY} AS c`,
      where: _where,
      values: _values,
    })
    if (count <= 0) throw new ServiceError('No data', { status: 204 })

    // get index
    let index = db.getIndex({
      table: _table,
      prefix: 'DISTINCT',
      field: _field,
      where: _where,
      join: _join,
      order: Boolean(query.order || query.sort) ? query.order : 'srl',
      sort: Boolean(query.order || query.sort) ? query.sort : 'desc',
      page: query.page,
      size: query.size,
      values: _values,
    })

    // set where for module
    _where = []
    if (query.q)
    {
      switch (query.module)
      {
        case helper.MODULE.NEST:
          _where.push(`AND (a.title LIKE "%${query.q}%" OR a.content LIKE "%${query.q}%")`)
          break
        case helper.MODULE.JSON:
          _where.push(`AND (j.name LIKE "%${query.q}%" OR j.description LIKE "%${query.q}%")`)
          break
      }
    }

    // set MOD
    const _mod: MOD = new MOD(query.mod)

    // transform items
    index.data = index.data.map((item: ZZ) => {
      if (!(item.srl && item.module)) return item
      // MOD / count
      if (_mod.check('count'))
      {
        let _itemWhere = [
          ..._where,
          `AND category_srl = ${item.srl}`,
        ]
        if (query.module && _tags) _itemWhere.push(getCountTagQuery(query.module, _tags))
        switch (item.module)
        {
          case helper.MODULE.NEST:
            item.count = articleHelper.count({
              table: `${DB.TABLE.ARTICLE} as a`,
              where: _itemWhere,
            })
            break
          case helper.MODULE.JSON:
            item.count = jsonHelper.count({
              table: `${DB.TABLE.JSON} as j`,
              where: _itemWhere,
            })
            break
        }
      }
      return item
    })

    // MOD / none
    if (_mod.check('none'))
    {
      let _newItem: ZZ = { name: 'none' }
      let _itemWhere = [ `AND category_srl IS NULL`, ..._where ]
      if (_mod.check('count'))
      {
        if (query.module && _tags) _itemWhere.push(getCountTagQuery(query.module, _tags))
        switch (query.module)
        {
          case helper.MODULE.NEST:
            if (query.module_srl) _itemWhere.push(`AND module_srl = ${query.module_srl}`)
            _newItem.count = articleHelper.count({
              table: `${DB.TABLE.ARTICLE} as a`,
              where: _itemWhere,
            })
            break
          case helper.MODULE.JSON:
            _newItem.count = jsonHelper.count({
              table: `${DB.TABLE.JSON} as j`,
              where: _itemWhere,
            })
            break
        }
      }
      index.data.push(_newItem)
    }

    // MOD / all
    if (_mod.check('all'))
    {
      let _newItem: ZZ = { name: 'all' }
      let _itemWhere = [ ..._where ]
      if (_mod.check('count'))
      {
        if (query.module && _tags) _itemWhere.push(getCountTagQuery(query.module, _tags))
        switch (query.module)
        {
          case helper.MODULE.NEST:
            if (query.module_srl) _itemWhere.push(`AND nest_srl = ${query.module_srl}`)
            _newItem.count = articleHelper.count({
              table: `${DB.TABLE.ARTICLE} as a`,
              where: _itemWhere,
            })
            break
          case helper.MODULE.JSON:
            _newItem.count = jsonHelper.count({
              table: `${DB.TABLE.JSON} as j`,
              where: _itemWhere,
            })
            break
        }
      }
      index.data.unshift(_newItem)
    }

    return {
      total: count,
      index: index.data,
    }
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to get Category index.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}

function getCountTagQuery(module: string, tags: string): string
{
  return `AND srl IN (SELECT mt.module_srl FROM ${DB.TABLE.MAP_TAG} AS mt WHERE mt.module LIKE \'${module}\' AND mt.tag_srl IN (${tags}))`
}
